odoo/tools/safe_eval.py

```diff
@@ -367,48 +376,35 @@ def safe_eval(expr, globals_dict=None, locals_dict=None, mode="eval", nocopy=Fal
     if type(expr) is CodeType:
         raise TypeError("safe_eval does not allow direct evaluation of code objects.")
 
-    # prevent altering the globals/locals from within the sandbox
-    # by taking a copy.
-    if not nocopy:
-        # isinstance() does not work below, we want *exactly* the dict class
-        if (globals_dict is not None and type(globals_dict) is not dict) \
-                or (locals_dict is not None and type(locals_dict) is not dict):
-            _logger.warning(
-                "Looks like you are trying to pass a dynamic environment, "
-                "you should probably pass nocopy=True to safe_eval().")
-        if globals_dict is not None:
-            globals_dict = dict(globals_dict)
-        if locals_dict is not None:
-            locals_dict = dict(locals_dict)
-
-    check_values(globals_dict)
-    check_values(locals_dict)
-
-    if globals_dict is None:
-        globals_dict = {}
-
-    globals_dict['__builtins__'] = dict(_BUILTINS)
-    if locals_builtins:
-        if locals_dict is None:
-            locals_dict = {}
-        locals_dict.update(_BUILTINS)
+    assert context is None or type(context) is dict, "Context must be a dict"
+
+    check_values(context)
+
+    globals_dict = dict(context or {}, __builtins__=dict(_BUILTINS))
+
     c = test_expr(expr, _SAFE_OPCODES, mode=mode, filename=filename)
     try:
-        return unsafe_eval(c, globals_dict, locals_dict)
-    except odoo.exceptions.UserError:
-        raise
-    except odoo.exceptions.RedirectWarning:
-        raise
-    except werkzeug.exceptions.HTTPException:
-        raise
-    except OperationalError:
-        # Do not hide PostgreSQL low-level exceptions, to let the auto-replay
-        # of serialized transactions work its magic
-        raise
-    except ZeroDivisionError:
+        # empty locals dict makes the eval behave like top-level code
+        return unsafe_eval(c, globals_dict, None)
+
+    except (
+        odoo.exceptions.UserError,
+        odoo.exceptions.RedirectWarning,
+        werkzeug.exceptions.HTTPException,
+        OperationalError,  # let auto-replay of serialized transactions work its magic
+        ZeroDivisionError,
+    ):
         raise
```

Comment:
Can you please extract this exception list (actually, a tuple) to a top-level variable, so it can be used in the `upgrade-util` PR?

```python
_BUBBLEUP_EXCEPTIONS = (
    odoo.exceptions.UserError,
    odoo.exceptions.RedirectWarning,
    werkzeug.exceptions.HTTPException,
    OperationalError,  # let auto-replay of serialized transactions work its magic
    ZeroDivisionError,
)
```

```suggestion
    except _BUBBLEUP_EXCEPTIONS:
        raise
```